nexus/src/app/image.rs

```diff
@@ -24,54 +25,77 @@ use omicron_common::api::external::LookupResult;
                 project: None,
             } => {
-                let image = LookupPath::new(opctx, &self.db_datastore)
-                    .image_id(id);
-                Ok(image)
+                let (.., db_image) = LookupPath::new(opctx, &self.db_datastore)
+                    .image_id(id).fetch().await?;
+                let lookup = match db_image.project_id {
+                    Some(id) => ImageLookup::ProjectImage(LookupPath::new(opctx, &self.db_datastore)
+                        .project_image_id(id)),
+                    None => {
+                        ImageLookup::SiloImage(LookupPath::new(opctx, &self.db_datastore)
+                            .silo_image_id(db_image.silo_id))
+                    },
+                };
+                Ok(lookup)
             }
             params::ImageSelector {
                 image: NameOrId::Name(name),
                 project: Some(project),
             } => {
                 let image =
-                    self.project_lookup(opctx, params::ProjectSelector { project })?.image_name_owned(name.into());
-                Ok(image)
+                    self.project_lookup(opctx, params::ProjectSelector { project })?.project_image_name_owned(name.into());
+                Ok(ImageLookup::ProjectImage(image))
+            }
+            params::ImageSelector {
+                image: NameOrId::Name(name),
+                project: None,
+            } => {
+                let image = self.current_silo_lookup(opctx)?.silo_image_name_owned(name.into());
+                Ok(ImageLookup::SiloImage(image))
             }
             params::ImageSelector {
                 image: NameOrId::Id(_),
                 ..
             } => Err(Error::invalid_request(
                 "when providing image as an ID, project should not be specified",
             )),
-            _ => Err(Error::invalid_request(
-                "image should either be UUID or project should be specified",
-            )),
             }
     }
 
-    /// Creates a project image
+    /// Creates an image
     pub async fn image_create(
         self: &Arc<Self>,
         opctx: &OpContext,
-        lookup_project: &lookup::Project<'_>,
+        lookup_parent: &ImageParentLookup<'_>,
         params: &params::ImageCreate,
     ) -> CreateResult<db::model::Image> {
-        let (.., authz_project) =
-            lookup_project.lookup_for(authz::Action::CreateChild).await?;
+        let (authz_silo, maybe_authz_project) = match lookup_parent {
+            ImageParentLookup::Project(project) => {
+                let (authz_silo, authz_project) =
+                    project.lookup_for(authz::Action::CreateChild).await?;
+                (authz_silo, Some(authz_project))
+            }
+            ImageParentLookup::Silo(silo) => {
+                let (.., authz_silo) =
+                    silo.lookup_for(authz::Action::CreateChild).await?;
+                (authz_silo, None)
+            }
+        };
```

Comment:
cute

Reply:
It helps avoid repeating the rather complex logic contained below.